core/corehttp/gateway_handler_car.go

```diff
@@ -0,0 +1,59 @@
+package corehttp
+
+import (
+	"context"
+	"net/http"
+
+	blocks "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	ipath "github.com/ipfs/interface-go-ipfs-core/path"
+	gocar "github.com/ipld/go-car"
+	selectorparse "github.com/ipld/go-ipld-prime/traversal/selector/parse"
+)
+
+// serveCar returns a CAR stream for specific DAG+selector
+func (i *gatewayHandler) serveCar(w http.ResponseWriter, r *http.Request, rootCid cid.Cid, contentPath ipath.Path) {
+	ctx := r.Context()
```

Comment:
```suggestion
	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()
```

This is because you are loosing gocar's context in the `dagstore` API wrapper:
```go
func (ds dagStore) Get(/* no context because of the API*/ c cid.Cid) (blocks.Block, error) {
```
Hopefully that will forward the cancel to dag API requests if the http write fail at some point.